rhel/repositoryscanner.go

```diff
@@ -204,49 +209,126 @@ func (r *RepositoryScanner) Scan(ctx context.Context, l *claircore.Layer) (repos
 	if !ok || cm == nil {
 		return []*claircore.Repository{}, fmt.Errorf("rhel: unable to create a mappingFile object")
 	}
-	CPEs, err := mapContentSets(ctx, sys, cm)
+
+	// First, look at DNF.
+	repoids, err := dnf.FindRepoids(ctx, sys)
+	switch {
+	case err != nil: // continue to error check and return
+	case len(repoids) == 0:
+		// If DNF yielded nothing, check for content-sets data.
+		repoids, err = repoidsFromContentSets(ctx, sys)
+	default:
+	}
 	if err != nil {
 		return []*claircore.Repository{}, err
 	}
-	if CPEs == nil && r.apiFetcher != nil {
-		// Embedded content-sets are available only for new images.
-		// For old images, use fallback option and query Red Hat Container API.
-		ctx, done := context.WithTimeout(ctx, r.cfg.Timeout)
-		defer done()
-		CPEs, err = mapContainerAPI(ctx, sys, r.apiFetcher)
+
+	pairs := func(yield func(string, string) bool) {
+		var found bool
+		for _, repoid := range repoids {
+			cpes, ok := cm.GetOne(ctx, repoid)
+			if !ok {
+				continue
+			}
+			found = true
+			for _, cpe := range cpes {
+				if !yield(repoid, cpe) {
+					return
+				}
+			}
+		}
+		if found {
+			return
+		}
+
+		if r.apiFetcher != nil {
+			// Embedded content-sets are unavailable in very old images.
+			// For these, use fallback option and query Red Hat Container API.
+			ctx, done := context.WithTimeout(ctx, r.cfg.Timeout)
+			defer done()
+
+			// !!! This ends up being a weird nonlocal return. !!!
+			var cpes []string
+			cpes, err = mapContainerAPI(ctx, sys, r.apiFetcher)
+			if err != nil {
+				return
+			}
+			for _, cpe := range cpes {
+				if !yield("", cpe) {
+					return
+				}
+			}
+		}
+	}
+	var repositories []*claircore.Repository
+	for repoid, cpeID := range pairs {
+		// TODO(hank) Remove these ContainerAPI hacks:
+		// Start ContainerAPI hacks
 		if err != nil {
 			return []*claircore.Repository{}, err
 		}
-	}
+		if repoid == "" {
+			repoid = cpeID
+		}
+		// End ContainerAPI hacks
 
-	for _, cpeID := range CPEs {
 		r := &claircore.Repository{
-			Name: cpeID,
+			Name: repoid,
 			Key:  repositoryKey,
 		}
 		r.CPE, err = cpe.Unbind(cpeID)
 		if err != nil {
 			zlog.Warn(ctx).
 				Err(err).
-				Str("url", `https://bugzilla.redhat.com/enter_bug.cgi?product=Container%20Factory`).
+				Str("url", bugURL(cpeID, err)).
 				Str("cpeID", cpeID).
 				Msg("invalid CPE, please report a bug upstream")
 			continue
 		}
 
 		repositories = append(repositories, r)
 	}
+	slices.SortFunc(repositories, func(a, b *claircore.Repository) int {
```

Comment:
can you explain the purpose of this?

Reply:
It removes duplicated repositories. This can apparently happen with the Rube Goldberg machine that is the build system.